### PR TITLE
US691069: QoQ Auto tune the target queue lengths based on average consumption rates

### PR DESCRIPTION
**Pull request by caoimhelynch**
https://internal.almoctane.com/ui/entity-navigation?p=131002/6001&entityType=work_item&id=691069

### CODE REVIEW

File: worker-message-prioritization-target-capacity-calculators/src/test/java/com/github/workerframework/workermessageprioritization/redistribution/TunedTargetQueueLengthProviderTest.java [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+                historicalConsumptionRate, roundTargetQueueLength, true, queueProcessingTimeGoalSeconds);
+
+        tunedTargetQueueLength = getTunedTargetQueueLength(targetQueue1,targetQueue);
+
+        assertEquals("NoOpMode is on: Target queue length should not have changed. Suggested adjustment for the target queue " +
+                "length should still be logged.", targetQueueLength, tunedTargetQueueLength);
+    }
+
+    @Test
+    public void getTunedTargetQueueForQueueWithInadequateHistoryAndNoOpModeOffTest(){
```
[the remark this answers is not in the document]
Comment (andyreidz) [reply worker-message-prioritization-target-capacity-calculators/src/test/java/com/github/workerframework/workermessageprioritization/redistribution/TunedTargetQueueLengthProviderTest.java:141]: What's highlighting this possible redundancy is that the asserts are the same for both tests, implying that the variables in parameters between both tests have no impact on the result.

image

image

File: worker-message-prioritization-target-capacity-calculators/src/test/java/com/github/workerframework/workermessageprioritization/redistribution/TunedTargetQueueLengthProviderTest.java [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+                historicalConsumptionRate, roundTargetQueueLength, true, queueProcessingTimeGoalSeconds);
+
+        tunedTargetQueueLength = getTunedTargetQueueLength(targetQueue1,targetQueue);
+
+        assertEquals("NoOpMode is on: Target queue length should not have changed. Suggested adjustment for the target queue" +
+                " length should still be logged.", targetQueueLength, tunedTargetQueueLength);
+    }
+
+    @Test
+    public void getMinTunedTargetQueueNoOpTest(){
```
[the remark this answers is not in the document]
Comment (andyreidz) [reply worker-message-prioritization-target-capacity-calculators/src/test/java/com/github/workerframework/workermessageprioritization/redistribution/TunedTargetQueueLengthProviderTest.java:110]: Will leave reviewing the rest of the tests until you have a chance to apply current feedback.

**Review by andyreidz — CHANGES_REQUESTED** [review]
Thanks Caoimhe, great work.

**Comment by andyreidz** [comment]
I forgot to highlight the format for logging statements.

image

The above avoids building the log message when the log level is disabled.

Also if you wanto still have the logs generated during build, set the CAF_LOG_LEVEL to DEBUG in the failsafe plugin configuration.

File: worker-message-prioritization-target-capacity-calculators/src/test/java/com/github/workerframework/workermessageprioritization/redistribution/HistoricalConsumptionRateTest.java [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    private final String targetQueue1 = "targetQueue1";
+    private final String targetQueue2 = "targetQueue2";
+
+    @Test
+    public void getErrorWhenMinConsumptionRateHistorySizeLargerThanMaxConsumptionRateHistorySizeTest() {
+
+        try {
+            final HistoricalConsumptionRate historicalConsumptionRate = new HistoricalConsumptionRate(10,
+                    100);
+            fail();
```
Comment (andyreidz) [inline worker-message-prioritization-target-capacity-calculators/src/test/java/com/github/workerframework/workermessageprioritization/redistribution/HistoricalConsumptionRateTest.java:42]:
```suggestion
            fail("IllegalArgumentException was not thrown");
```

File: worker-message-prioritization-target-capacity-calculators/src/test/java/com/github/workerframework/workermessageprioritization/redistribution/HistoricalConsumptionRateTest.java [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+        } catch (final IllegalArgumentException exception) {
+            assertNotNull("IllegalArgumentException was not thrown", exception);
```
Comment (andyreidz) [inline worker-message-prioritization-target-capacity-calculators/src/test/java/com/github/workerframework/workermessageprioritization/redistribution/HistoricalConsumptionRateTest.java:44]:
```suggestion
```
exception can't be null here

Comment (andyreidz) [reply worker-message-prioritization-target-capacity-calculators/src/test/java/com/github/workerframework/workermessageprioritization/redistribution/HistoricalConsumptionRateTest.java:44]: Apply the same change where required.

File: worker-message-prioritization-target-capacity-calculators/src/test/java/com/github/workerframework/workermessageprioritization/redistribution/HistoricalConsumptionRateTest.java [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+        final HistoricalConsumptionRate historicalConsumptionRate = new HistoricalConsumptionRate(maximumConsumptionRateHistorySize,
+                minConsumptionRateHistorySize);
+
+        IntStream.range(0, minConsumptionRateHistorySize).forEach(i ->
+            historicalConsumptionRate.recordCurrentConsumptionRateHistoryAndGetAverage(targetQueue1, theoreticalConsumptionRate1));
+
+        IntStream.range(0, minConsumptionRateHistorySize + 5).forEach(i ->
+            historicalConsumptionRate.recordCurrentConsumptionRateHistoryAndGetAverage(targetQueue2, theoreticalConsumptionRate2));
+
+        final boolean theoreticalConsumptionRateHistoryQueue1 = historicalConsumptionRate.isSufficientHistoryAvailable(targetQueue1);
```
Comment (andyreidz) [inline worker-message-prioritization-target-capacity-calculators/src/test/java/com/github/workerframework/workermessageprioritization/redistribution/HistoricalConsumptionRateTest.java:80]:
```suggestion
        final boolean isSufficientHistoryAvailableQueue1 = historicalConsumptionRate.isSufficientHistoryAvailable(targetQueue1);
```

File: worker-message-prioritization-target-capacity-calculators/src/test/java/com/github/workerframework/workermessageprioritization/redistribution/HistoricalConsumptionRateTest.java [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+        final boolean theoreticalConsumptionRateHistoryQueue2 = historicalConsumptionRate.isSufficientHistoryAvailable(targetQueue2);
```
Comment (andyreidz) [inline worker-message-prioritization-target-capacity-calculators/src/test/java/com/github/workerframework/workermessageprioritization/redistribution/HistoricalConsumptionRateTest.java:82]:
```suggestion
        final boolean isSufficientHistoryAvailableQueue2 = historicalConsumptionRate.isSufficientHistoryAvailable(targetQueue2);
```

File: worker-message-prioritization-target-capacity-calculators/src/test/java/com/github/workerframework/workermessageprioritization/redistribution/RoundTargetQueueLengthTest.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertEquals;
+
+public class RoundTargetQueueLengthTest {
+
+    @Rule
+    public final ErrorCollector collector = new ErrorCollector();
+
+    @Test
+    public void getRoundedTargetQueueLengthTest() {
```
Comment (andyreidz) [inline worker-message-prioritization-target-capacity-calculators/src/test/java/com/github/workerframework/workermessageprioritization/redistribution/RoundTargetQueueLengthTest.java:35]: Minor one, spaces after commas 👇

File: worker-message-prioritization-target-capacity-calculators/src/test/java/com/github/workerframework/workermessageprioritization/redistribution/RoundTargetQueueLengthTest.java [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+
+    @Test
+    public void roundingMultipleCannotBeSetToZeroTest() {
+
+        try {
+            final RoundTargetQueueLength roundTargetQueueLength = new RoundTargetQueueLength(0);
+            fail();
+        } catch (final IllegalArgumentException exception){
+            assertNotNull("IllegalArgumentException was not thrown", exception);
```
Comment (andyreidz) [inline worker-message-prioritization-target-capacity-calculators/src/test/java/com/github/workerframework/workermessageprioritization/redistribution/RoundTargetQueueLengthTest.java:79]: Apply previous comment use "IllegalArgumentException was not thrown" in the fail() and remove the assertNotNull

File: worker-message-prioritization-target-capacity-calculators/src/test/java/com/github/workerframework/workermessageprioritization/redistribution/TunedTargetQueueLengthProviderTest.java [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+        final TargetQueuePerformanceMetricsProvider targetQueuePerformanceMetricsProvider =
+                mock(TargetQueuePerformanceMetricsProvider.class);
+        when(targetQueuePerformanceMetricsProvider.getTargetQueuePerformanceMetrics(targetQueue1)).thenReturn(performanceMetrics);
+
+        final TunedTargetQueueLengthProvider targetQueue = new TunedTargetQueueLengthProvider(targetQueuePerformanceMetricsProvider,
+                historicalConsumptionRate, roundTargetQueueLength, true, queueProcessingTimeGoalSeconds);
+
+        final long tunedTargetQueueLength = getTunedTargetQueueLength(targetQueue1,targetQueue);
+
+        assertEquals("NoOpMode is on: Target queue length should not have changed.Suggested adjustment for the target queue length " +
```
Comment (andyreidz) [inline worker-message-prioritization-target-capacity-calculators/src/test/java/com/github/workerframework/workermessageprioritization/redistribution/TunedTargetQueueLengthProviderTest.java:71]:
```suggestion
        assertEquals("NoOpMode is on: Target queue length should not have changed. Suggested adjustment for the target queue length " +
```

File: worker-message-prioritization-target-capacity-calculators/src/test/java/com/github/workerframework/workermessageprioritization/redistribution/TunedTargetQueueLengthProviderTest.java [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+        when(roundTargetQueueLength.getRoundedTargetQueueLength(ArgumentMatchers.anyLong())).thenReturn(300L);
+
+        final TargetQueuePerformanceMetricsProvider targetQueuePerformanceMetricsProvider =
+                mock(TargetQueuePerformanceMetricsProvider.class);
+        when(targetQueuePerformanceMetricsProvider.getTargetQueuePerformanceMetrics(targetQueue1)).thenReturn(performanceMetrics);
+
+        final TunedTargetQueueLengthProvider targetQueue = new TunedTargetQueueLengthProvider(targetQueuePerformanceMetricsProvider,
+                historicalConsumptionRate, roundTargetQueueLength, false, queueProcessingTimeGoalSeconds);
+
+        final long tunedTargetQueueLength = getTunedTargetQueueLength(targetQueue1,targetQueue);
```
Comment (andyreidz) [inline worker-message-prioritization-target-capacity-calculators/src/test/java/com/github/workerframework/workermessageprioritization/redistribution/TunedTargetQueueLengthProviderTest.java:99]:
```suggestion
        final long tunedTargetQueueLength = getTunedTargetQueueLength(targetQueue1, targetQueue);
```

**Review by andyreidz — CHANGES_REQUESTED** [review]
Looks good, some small comments about formatting and checking for exceptions in tests.

Can you also move to using this style of logging?

LOGGER.debug("Target queue length has been rounded from: " + tunedTargetQueue + " to:" + roundedQueueLength);

LOGGER.debug("Target queue length has been rounded from: {} to {}", tunedTargetQueue , roundedQueueLength);

File: worker-message-prioritization-distribution/src/test/java/com/github/workerframework/workermessageprioritization/redistribution/consumption/MinimumConsumptionTargetCalculatorTest.java [hunk 11]
```diff
@@ -37,16 +45,35 @@ public final class MinimumConsumptionTargetCalculatorTest
...
     public void getTargetQueueCapacityTest()
     {
-        final TargetQueueSettingsProvider provider = mock(TargetQueueSettingsProvider.class);
+        final TargetQueueSettingsProvider targetQueueSettingsProvider = mock(TargetQueueSettingsProvider.class);
         final Queue targetQueue = new Queue();
         targetQueue.setMessages(750);
 
-        final TargetQueueSettings settings = new TargetQueueSettings(1000, 20);
-        when(provider.get(targetQueue)).thenReturn(settings);
+        final TargetQueueSettings settings = new TargetQueueSettings(1000, 20, 1, 1);
```
[the remark this answers is not in the document]
Comment (caoimhelynch) [reply worker-message-prioritization-distribution/src/test/java/com/github/workerframework/workermessageprioritization/redistribution/consumption/MinimumConsumptionTargetCalculatorTest.java:52]: Have altered these to make the different in the tests more clear.

**Pull request closed without merging**